### DISCUSSION TITLE
The difference of Partition and Partition Key

### DISCUSSION THREAD

**Discussion by bingbb47** [milvus-io/milvus#26320] [Q&A and General discussion] [answered]
When create a partition，will produce [shard_nums] segments. This leads to a large number of segments generated when a collection contains many partitions, which may slow down operations such as load.
According to my understanding, the Partition Key should have avoided this.
Are the two equivalent in terms of performance? Is Partition key a more recommended approach now？

**Comment by xiaofan-luan** [comment, marked as answer]
Partition is more like physical isolation, where each partition will be in one segment.
Which means partition number can not be huge, that's why milvus has a upper limit of 4096 partitions.

While partition key is a logical isolation, each partition key is hashed into a parititon(usually partition is a fixed number, for instance 64 partitions). While search, if partition is given, then we first find the valid physical partition, then we do filtering on this segments

**Reply by bingbb47** [reply]
Can I assume that the Partition and Partition key  both do the same thing, and that the Partition key cannot be used to create too many partitions too

**Reply by xiaofan-luan** [reply]
Exactly true

**Reply by ShravanSunder** [reply]
@xiaofan-luan If we use userId as a partition key, that would be an anti patter due to the upper limit?
The docs seem to say we there is no upper bound on the number of partition keys.  https://milvus.io/docs/multi_tenancy.md#Partition-key-based-multi-tenancy

**Reply by lyajedi** [reply]
@xiaofan-luan 
I have another question; will a partition contain multiple different values of a partition key?

I have a partition key "year_month" with YYYYMM as value, what if I conduct a search with expression "year_month=202309", will Milvus only load a single partition instead of the whole collection?

Also, only 64 partitions created (named as _default_0-63) after the collection created with the partition key, what happens if I have more than 64 YYYMM values?

**Reply by yhmo** [reply]
@ShravanSunder @lyajedi 

The doc says no upper bound for the max number of tenants if you apply partition key. But the number of partitions is limited to 4096.
Assuming I have 1 million tenants(t0, t2, t3 .... t999999), I create a collection with a partition key and set the partition number to 4096. I insert the 1 million tenants into the collection, milvus automatically distributes them into different partitions: t0, t64, t128, t256...are sent to partition_0; t1, t65, t129, t257... are sent to partition_1; t2, t66, t130, t258... are sent to partition_2, etc.

We can see there are only 4096 partitions to handle the 1 million tenants.
"partition key" is a special field to split the data, the collection has only one "partition key". By hashing the values of this field, we can split data into different partitions.

For collection with partition key, collection.load() will load the entire collection. Not allowed to specify a single partition to load.

Let's say a collection has a partition key "year_month" with YYYMM values. And you have specified the partition_num=64. Definitely you can insert more than 64 YYYMM values. All the values are always hashed to the 64 partitions.

**Reply by lyajedi** [reply]
Thank you @yhmo 
I try to figure out which approach is the best to design my collection.

I have huge amount of data every month and I'd like to load a partition instead of the whole collection into memory as I have limited RAM resources allocated. 
My use case for conducting the search is also on a specific month, so if with the YYYYMM as the partition key, it seems I cannot just load a single partition as I don't know the partition name for entities with year_month=202309.
Should I just manually create the partitions with YYYYMM as partition name instead and insert the data into the corresponding partitions?

**Reply by ShravanSunder** [reply]
@yhmo How does `For collection with partition key, collection.load() will load the entire collection. Not allowed to specify a single partition to load.` impact a  Cost-optimized CU instance (mmap) on zillz?  Would this cause any issues if i have my data partition by userId.

Another question, do additional filter clauses reduce the memory needs of a query or is the whole collection still loaded due to the partition key of userid.

thanks :)

**Comment by xiaofan-luan** [comment]
> 



> There are no limit for partition key numbers.
Currently we have no api to delete by partition keys, and it would be published recently.
